test/hotspot/gtest/nmt/test_vmatree.cpp

```diff
@@ -0,0 +1,516 @@
+  struct Info {
+    Tpe tpe;
+    MEMFLAGS flag;
+    NativeCallStack stack;
+    Info() : tpe(Free), flag(mtNone), stack() {}
+
+    Info(Tpe tpe, NativeCallStack stack, MEMFLAGS flag)
+    : tpe(tpe), flag(flag), stack(stack) {}
+
+    bool eq(Info other) {
+      return flag == other.flag && stack.equals(other.stack);
+    }
+  };
+  // Page (4KiB) granular array
+  const size_t num_pages = 1024 * 512;
+  Info pages[1024 * 512];
+
+  SimpleVMATracker()
+  : pages() {
+    for (size_t i = 0; i < num_pages; i++) {
+      pages[i] = Info();
+    }
+  }
+
+  VMATree::SummaryDiff do_it(Tpe tpe, size_t start, size_t size, NativeCallStack stack, MEMFLAGS flag) {
+    assert(size % page_size == 0 && start % page_size == 0, "page alignment");
+    VMATree::SummaryDiff diff;
+    const size_t page_count = size / page_size;
+    const size_t start_idx = start / page_size;
+    const size_t end_idx = start_idx + page_count;
+    assert(end_idx < (size_t)num_pages, "");
+
+    Info new_info(tpe, stack, flag);
+    for (size_t i = start_idx; i < end_idx; i++) {
+      Info& old_info = pages[i];
+
+      // Register diff
+      if (old_info.tpe == Reserved) {
+        diff.flag[(int)old_info.flag].reserve -= page_size;
+      } else if (old_info.tpe == Committed) {
+        diff.flag[(int)old_info.flag].reserve -= page_size;
+        diff.flag[(int)old_info.flag].commit -= page_size;
+      }
+
+      if (tpe == Reserved) {
+        diff.flag[(int)new_info.flag].reserve += page_size;
+      } else if(tpe == Committed) {
+        diff.flag[(int)new_info.flag].reserve += page_size;
+        diff.flag[(int)new_info.flag].commit += page_size;
+      }
+      // Overwrite old one with new
+      pages[i] = new_info;
+    }
+    return diff;
+  }
+
+  VMATree::SummaryDiff reserve(size_t start, size_t size, NativeCallStack stack, MEMFLAGS flag) {
+    return do_it(Reserved, start, size, stack, flag);
+  }
+
+  VMATree::SummaryDiff commit(size_t start, size_t size, NativeCallStack stack, MEMFLAGS flag) {
+    return do_it(Committed, start, size, stack, flag);
+  }
+
+  VMATree::SummaryDiff release(size_t start, size_t size) {
+    return do_it(Free, start, size, NativeCallStack(), mtNone);
+  }
+};
+
+TEST_VM_F(VMATreeTest, TestConsistencyWithSimpleTracker) {
+  // In this test we use ASSERT macros from gtest instead of EXPECT
+  // as any error will propagate and become larger as the test progresses.
+  SimpleVMATracker* tr = new SimpleVMATracker();
+  const size_t page_size = tr->page_size;
+  VMATree tree;
+  NativeCallStackStorage ncss(true);
+
+  NativeCallStack stacks[4] = {
+    make_stack(0xA),
+    make_stack(0xB),
+    make_stack(0xC),
+    make_stack(0xD)
+  };
+
+
+  const int operation_count = 100000; // One hundred thousand
+  for (int i = 0; i < operation_count; i++) {
+    const size_t page_start = (size_t)(os::random() % tr->num_pages);
+    const size_t num_pages = (size_t)(os::random() % (tr->num_pages - page_start));
```

Comment:
Proposal: roll dice for two positions in num_pages range. Then sort them (swap if A > B). 

That gives you a fairer distribution, since you don't overemphasize the end of the range.